libmcount/dynamic.c

```diff
@@ -327,10 +351,17 @@ static void prepare_dynamic_update(struct uftrace_sym_info *sinfo, bool needs_mo
 	};
 	int hash_size = sinfo->exec_map->mod->symtab.nr_sym * 3 / 4;
 
-	if (needs_modules)
-		hash_size *= 2;
-
-	code_hmap = hashmap_create(hash_size, hashmap_ptr_hash, hashmap_ptr_equals);
+	if (code_hmap && !needs_modules) /* already prepared */
+		return;
+	if (!code_hmap) { /* nothing loaded */
+		if (needs_modules)
+			hash_size *= 2;
+		code_hmap = hashmap_create(hash_size, hashmap_ptr_hash, hashmap_ptr_equals);
+	}
+	else if (mdinfo && !mdinfo->next) /* main binary loaded, no module loaded */
+		fmd.main_loaded = true;
+	else
+		return;
```

Comment:
This condition check is confusing as it sometimes checks `code_hmap` and sometimes `mdinfo`.  I think you can just call `dl_iterate_phdr()` always and check the `mdinfo` list inside to prevent duplication.

Reply:
There are performance implications if we traverse the mdinfo linked list every time we want check for a shared library. Therefore, I decided to add a new static variable 'modules_loaded' which takes this into account and will not run callback if modules are already loaded.